src/vcf/report_writer.cpp

```diff
@@ -0,0 +1,165 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "vcf/report_writer.hpp"
+
+namespace ebi
+{
+  namespace vcf
+  {
+    SqliteReportWriter::SqliteReportWriter(std::string db_name) : db_name{db_name}, current_transaction_size{0},
+                                                      transaction_size{1000000}, sleep_time{500}
+    {
+        int rc = sqlite3_open(db_name.c_str(), &db);
+        if (rc != SQLITE_OK) {
+            sqlite3_close(db);
+            throw std::runtime_error(std::string("Can't open database: ") + sqlite3_errmsg(db));
+        }
+
+        char *zErrMsg = NULL;
+        rc = sqlite3_exec(db, "CREATE TABLE if not exists errors (line int, message varchar(255));", NULL, 0,
+                          &zErrMsg);
+        if (rc != SQLITE_OK) {
+            std::string error_message{std::string("Can't use table: ") + zErrMsg};
+            sqlite3_free(zErrMsg);
+            sqlite3_close(db);
+            throw std::runtime_error{error_message};
+        }
+        
+        rc = sqlite3_exec(db, "CREATE TABLE if not exists warnings (line int, message varchar(255));", NULL, 0,
+                          &zErrMsg);
+        if (rc != SQLITE_OK) {
+            std::string error_message{std::string("Can't use table: ") + zErrMsg};
+            sqlite3_free(zErrMsg);
+            sqlite3_close(db);
+            throw std::runtime_error{error_message};
+        }
+    }
+
+    void SqliteReportWriter::close()
+    {
+        // in case the amount of inserts is not a multiple of the transaction size, do the leftover inserts in a smaller transaction
+        if (current_transaction_size != 0) {
+            commit_transaction();
+            current_transaction_size = 0;
+        }
+
+        if (db != nullptr) {
+            int rc = sqlite3_close(db);
+            if (rc != SQLITE_OK) {
+                int remaining_tries = 10;
+                while (rc == SQLITE_BUSY && remaining_tries-- > 0) {
+                    // maybe sqlite is still writing the last transaction
+                    std::cerr << "waiting to finish transactions" << std::endl;
+                    std::this_thread::sleep_for(sleep_time);
+                    rc = sqlite3_close(db);
+                }
+                if (rc != SQLITE_OK) {
+                    throw std::runtime_error{"can not close database " + db_name};
+                }
+            }
+            db = nullptr;
+        }
+    }
+
+    SqliteReportWriter::~SqliteReportWriter()
+    {
+        try {
+            close();
+        } catch (std::runtime_error e) {
+            std::cerr << "Error at ~SqliteReportRW: " << e.what() << std::endl;
```

Comment:
Maybe we could try to make this error clearer when notified to the user? Something like "An error occurred when finalizing the report generation", using a pun with the meaning of "finalizing".